components/ca-form.js

```diff
@@ -0,0 +1,1520 @@
+                        for (const key in createSchema) {
+
+                            if (Object.prototype.hasOwnProperty.call(createSchema, key) && this.createUrl.indexOf(`{${key}}`)) {
+
+                                const schemaItem = createSchema[key];
+
+                                // if we have a required guid, add one
+                                if (schemaItem.required && schemaItem.format === 'uuid') {
+                                    this.createUrl = this.createUrl.replace(`{${key}}`, newGuid());
+                                }
+                            }
+                        }
+                    }
+
+                    // insert button container
+                    buttonContainer = createElementLegacy(this.form, 'div', { class: 'buttons' });
+
+                    // add submit button
+                    createElementLegacy(buttonContainer, 'input', { type: 'submit', value: createLink.title || 'Save' });
+
+                    // add reset button if we have it.
+                    if (this.reset) {
+                        resetButton = createElementLegacy(buttonContainer, 'input', { name: 'reset', type: 'button', value: 'Reset' });
+                        resetButton.onclick = () => {
+                            this.src = this.src;
+                        };
+                    }
+                }
+
+                // add buttons from array if we have them
+                if (this.buttons.length > 0) {
+
+                    if (!buttonContainer) {
+
+                        // insert button container if it does not already exist
+                        buttonContainer = createElementLegacy(this.form, 'div', { class: 'buttons' });
+                    }
+
+                    // add a button for each item in the array
+                    for (let i = 0, l = this.buttons.length; i < l; i++) {
+
+                        // insert button
+                        const button = createElementLegacy(buttonContainer, 'input', { type: 'button', value: this.buttons[i] });
+
+                        // bubble click event passing button display value to external handler
+                        button.onclick = (e) => {
+                            cancelEvent(e);
+                            this.onbuttonclick(this, button.value);
+                        };
+                    }
+                }
+
+                // fire onrendercomplete providing all child elements have finished rendering
+                this.checkRenderComplete();
+            }
+        }
+
+        /**
+         * @description Goes through all input elements and checks if a rendered property exists and is true
+         * if not, checks again every 250ms eventually firing the onrendercomplete event once all children are considered rendered
+         * @returns {Boolean} checks all properties rendered.
+         */
+        checkRenderComplete() {
+
+            const unRenderedElements = this.elements.filter(item =>
```

Comment:
the capitalisation of un(R)endered is weird... should be unrenderedElements (imho)